src/compiler/checker.ts

```diff
@@ -1672,13 +1672,17 @@ namespace ts {
                 return isForInOrOfStatement(grandparent) && isSameScopeDescendentOf(usage, grandparent.expression, declContainer);
             }
 
-            function isUsedInFunctionOrInstanceProperty(usage: Node, declaration: Node): boolean {
+            function isUsedInFunctionOrInstanceProperty(usage: Node, declaration: Node) {
+                return isUsedInFunctionOrInstancePropertyWorker(usage, declaration);
+            }
+
+            function isUsedInFunctionOrInstancePropertyWorker(usage: Node, declaration: Node): boolean {
                 return !!findAncestor(usage, current => {
                     if (current === declContainer) {
                         return "quit";
                     }
                     if (isFunctionLike(current)) {
-                        return true;
+                        return !getImmediatelyInvokedFunctionExpression(current);

```

Comment:
Enum is not checked properly.  [playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcCMB2AFASkQbwFgAoRMxAGwFMpEBDRAXkS1yYD5EBRMEAWwB0ATSoBnHLgDcZAPQzEogBZwQFACaIqAJy1wtJcpt59uxggcPkAcnCaIADABoLlxCNF20z0uQC+Jf2ISFHRsHCA)

```ts
function foo17() {
    let a = (() => Enum.Yes)();  // should error
    enum Enum {
        No = 0,
        Yes = 1,
    }
}

foo17()
```